### PR TITLE
Chapter 2 - tasks completed

### PR DESCRIPTION
**Pull request by lucasDS-0**
### Solutions for Chapter 2

That's for ch2, any advice on the code is appreciated!

cc @vrom911 @chshersh

### CODE REVIEW

**Review by vrom911 — APPROVED** [review]
Great job with lists Chapter 🚆 👏🏼

File: src/Chapter2.hs [hunk 1]
```diff
@@ -620,8 +640,13 @@ Write a function that takes elements of a list only in even positions.
 
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
+
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven []     = []
+takeEven [x]    = [x]
+takeEven (x:xs) = x : takeEven (tail xs)
```
[the remark this answers is not in the document]
Comment (lucasDS-0) [reply src/Chapter2.hs:648]: That makes sense, I completely forgot about safety, thanks for pointing it out!

Is this implementation better?

```haskell
takeEven :: [a] -> [a]
takeEven []       = []
takeEven [x]      = [x]
takeEven (x:_:xs) = x : takeEven xs
```

File: src/Chapter2.hs [hunk 2]
```diff
@@ -727,8 +754,12 @@ value of the element itself
 
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
+
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l
+    | null l    = []
+    | otherwise = replicate (head l) (head l) ++ smartReplicate (tail l)
```
[the remark this answers is not in the document]
Comment (lucasDS-0) [reply src/Chapter2.hs:761]: I had to do a little bit of remembering for improving this one, I better review my notes!

Is this implementation better?

```haskell
smartReplicate :: [Int] -> [Int]
smartReplicate = foldr (\x -> (++) $ replicate x x) []
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -837,12 +872,17 @@ list.
...
 [4,1,2,3]
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+
+rotate :: Int -> [a] -> [a]
+rotate = \x l ->  if x < 0
+                  then []
+                  else take (length l) $ drop x $ cycle l
```
[the remark this answers is not in the document]
Comment (lucasDS-0) [reply src/Chapter2.hs:885]: I hadn't thought about that! Would this suffice?

```haskell
rotate :: Int -> [a] -> [a]
rotate x l =  
    if x < 0
    then []
    else take (length l) $ drop r $ cycle l
  where
    r = x `mod` length l
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -858,7 +898,11 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+
+rewind :: [a] -> [a]
+rewind = \l ->  if null l
+                then []
+                else rewind (tail l) ++ [head l]
```
[the remark this answers is not in the document]
Comment (lucasDS-0) [reply src/Chapter2.hs:905]: True, I completely overlooked it.

How about this implementation?

```haskell
rewind :: [a] -> [a]
rewind = go []
  where
    go :: [a] -> [a] -> [a]
    go acc []     = acc
    go acc (x:xs) = go (x:acc) xs
```

**Comment by lucasDS-0** [comment]
I just pushed this changes to the branch, let me know how these changes look like when you can!

File: src/Chapter2.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Chapter2.hs]: Perfect 👌🏼 💎

File: src/Chapter2.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Chapter2.hs]: That looks good.
But also, you can say the following:

```haskell
smartReplicate = concatMap (\x -> replicate x x)
```

File: src/Chapter2.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Chapter2.hs]: Yes, this is awesome!

**Comment by vrom911** [comment]
@lucasDS-0 this chapter looks fantastic now! Feel free to merge 🚢

File: src/Chapter2.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucasDS-0) [reply src/Chapter2.hs]: Good idea! I completely forgot about concatMap. I will change it to match this implementation.

**Comment by lucasDS-0** [comment]
Cool! To Chapter 3 then!

**Pull request merged**